src/jit/lower.cpp

```diff
@@ -1863,6 +1872,135 @@ GenTree* Lowering::LowerTailCallViaHelper(GenTreeCall* call, GenTree* callTarget
     return result;
 }
 
+void Lowering::LowerCompare(GenTree* cmp)
+{
+#ifndef _TARGET_64BIT_
+    if (cmp->gtGetOp1()->TypeGet() != TYP_LONG)
+    {
+        return;
+    }
+
+    LIR::Use cmpUse;
+    
+    if (!BlockRange().TryGetUse(cmp, &cmpUse) || cmpUse.User()->OperGet() != GT_JTRUE)
+    {
+        return;
+    }
+
+    GenTree* src1   = cmp->gtGetOp1();
+    GenTree* src2   = cmp->gtGetOp2();
+    unsigned weight = m_block->getBBWeight(comp);
+
+    LIR::Use loSrc1(BlockRange(), &(src1->gtOp.gtOp1), src1);
+    LIR::Use loSrc2(BlockRange(), &(src2->gtOp.gtOp1), src2);
+
+    if (loSrc1.Def()->OperGet() != GT_CNS_INT && loSrc1.Def()->OperGet() != GT_LCL_VAR)
```

Comment:
Note that this check is rather conservative. We should be able to also move `GT_IND` nodes but that's obviously more complicated because we'd need to move the address tree as well.

For example, comparing two static variables generates this code:
```
       8D053046EB00 lea      eax, [classVar[0xeb4e40]]
       8B10         mov      edx, dword ptr [eax]
       8B4004       mov      eax, dword ptr [eax+4]
       8D0D3846EB00 lea      ecx, [classVar[0xeb4e54]]
       8B31         mov      esi, dword ptr [ecx]
       8B4904       mov      ecx, dword ptr [ecx+4]
       3BC1         cmp      eax, ecx
       750A         jne      SHORT G_M55886_IG03
       3BD6         cmp      edx, esi
       7506         jne      SHORT G_M55886_IG03
       FF152C69EB00 call     [System.Console:WriteLine()]
```
Of course, this isn't compare specific, the problem actually starts in decomposition.